src/workerd/server/bundle-fs.c++

```diff
@@ -0,0 +1,433 @@
+              KJ_CASE_ONEOF(file, kj::Own<FileEntry>) {
+                // We have a file entry. This is not allowed. We will
+                // just ignore it.
+                break;
+              }
+            }
+          }
+        }
+
+        break;
+      }
+      case config::Worker::Which::SERVICE_WORKER_SCRIPT: {
+        auto locked = current->entries.lockExclusive();
+        // In this case there's only one file. We'll call it "worker"
+        locked->insert(kj::str("worker"), kj::atomicRefcounted<FileEntry>(*inner));
+        break;
+      }
+      case config::Worker::Which::INHERIT: {
+        // TODO(node-fs): The inherit case is not supported yet simply
+        // because of the additional complexity it entails. The idea is
+        // that this worker would inherit configuration from a different
+        // configuration, which assumes that this workers file system
+        // would inherit entries also. It's not at all clear just how
+        // critical this is, so for now we just ignore it.
+        return;
+      }
+    }
+  }
+
+  kj::Own<const DirectoryDelegate> clone() const override {
+    return kj::atomicAddRef(*this);
+  }
+
+  bool exists(kj::PathPtr path) const override {
+    return root->exists(path);
+  }
+
+  kj::Array<kj::String> listNames() const override {
+    return root->listNames();
+  }
+
+  kj::Array<kj::ReadableDirectory::Entry> listEntries() const override {
+    return root->listEntries();
+  }
+
+  kj::Maybe<kj::FsNode::Metadata> tryLstat(kj::PathPtr path) const override {
+    return root->tryLstat(path);
+  }
+
+  kj::Maybe<kj::Own<const kj::ReadableFile>> tryOpenFile(kj::PathPtr path) const override {
+    return root->tryOpenFile(path);
+  }
+
+  kj::Maybe<kj::Own<const kj::ReadableDirectory>> tryOpenSubdir(kj::PathPtr path) const override {
+    return root->tryOpenSubdir(path);
+  }
+
+ private:
+  kj::Own<config::Worker::Reader> inner;
+
+  struct DirectoryEntry;
+  struct FileEntry;
+
+  using ModuleEntry = kj::OneOf<kj::Own<DirectoryEntry>, kj::Own<FileEntry>>;
+
+  struct DirectoryEntry final: public kj::ReadableDirectory, public kj::AtomicRefcounted {
+    kj::MutexGuarded<kj::HashMap<kj::String, ModuleEntry>> entries;
+
+    kj::FsNode::Metadata stat() const override {
+      return kj::FsNode::Metadata{
+        .type = kj::FsNode::Type::DIRECTORY,
+        .size = 0,
+        .spaceUsed = 0,
+        .lastModified = kj::UNIX_EPOCH,
+        .linkCount = 1,
+        .hashCode = 0,
+      };
+    }
+
+    void sync() const override {}
+    void datasync() const override {}
+
+    kj::Own<const DirectoryEntry> addRef() const {
+      return kj::atomicAddRef(*this);
+    }
+
+    kj::Own<const kj::FsNode> cloneFsNode() const override {
+      return kj::atomicAddRef(*this);
+    }
+
+    bool exists(kj::PathPtr path) const override {
+      auto locked = entries.lockShared();
+      if (path.size() == 0) return true;
+      KJ_IF_SOME(found, locked->find(path[0])) {
```

Comment:
This path finding logic is duplicated in a lot of different places with only small modifications.
I think this can be generalized in a private method that just returns a Maybe<OneOf<FileEntry,DirectoryEntry>> or OneOf<void, FileEntry, DirectoryEntry> and each of these methods can use it and act appropriately  based on the result.

Reply:
Yep, already in plan. Will be done before this is moved out of draft.